src/ncp/ncp_base_mtd.cpp

```diff
@@ -818,6 +823,139 @@ template <> otError NcpBase::HandlePropertyRemove<SPINEL_PROP_THREAD_ON_MESH_NET
 }
 #endif // OPENTHREAD_ENABLE_BORDER_ROUTER
 
+#if OPENTHREAD_ENABLE_SERVICE
+
+template <> otError NcpBase::HandlePropertyGet<SPINEL_PROP_SERVER_ALLOW_LOCAL_DATA_CHANGE>(void)
+{
+    return mEncoder.WriteBool(mAllowLocalServerDataChange);
+}
+
+template <> otError NcpBase::HandlePropertySet<SPINEL_PROP_SERVER_ALLOW_LOCAL_DATA_CHANGE>(void)
+{
+    bool    value                    = false;
+    otError error                    = OT_ERROR_NONE;
+    bool    shouldRegisterWithLeader = false;
+
+    SuccessOrExit(error = mDecoder.ReadBool(value));
+
+    // Register any server data changes on transition from `true` to `false`.
+    shouldRegisterWithLeader = (mAllowLocalServerDataChange == true) && (value == false);
+
+    mAllowLocalServerDataChange = value;
+
+exit:
+
+    if (shouldRegisterWithLeader)
+    {
+        otServerRegister(mInstance);
+    }
+
+    return error;
+}
+
+template <> otError NcpBase::HandlePropertyInsert<SPINEL_PROP_SERVER_SERVICES>(void)
+{
+    otError         error = OT_ERROR_NONE;
+    otServiceConfig cfg;
+    bool            stable;
+    const uint8_t * data;
+    uint16_t        dataLen;
+
+    VerifyOrExit(mAllowLocalServerDataChange == true, error = OT_ERROR_INVALID_STATE);
+
+    SuccessOrExit(error = mDecoder.ReadUint32(cfg.mEnterpriseNumber));
+    SuccessOrExit(error = mDecoder.ReadDataWithLen(data, dataLen));
+
+    VerifyOrExit((dataLen <= sizeof(cfg.mServiceData) && dataLen <= UINT8_MAX), error = OT_ERROR_INVALID_ARGS);
+
+    memcpy(cfg.mServiceData, data, dataLen);
+    cfg.mServiceDataLength = static_cast<uint8_t>(dataLen);
+
+    SuccessOrExit(error = mDecoder.ReadBool(stable));
+    cfg.mServerConfig.mStable = stable;
+    SuccessOrExit(error = mDecoder.ReadDataWithLen(data, dataLen));
+
+    VerifyOrExit((dataLen <= sizeof(cfg.mServerConfig.mServerData) && dataLen <= UINT8_MAX),
```

Comment:
Same here. Doesn't `dataLen <= sizeof(cfg.mServerConfig.mServerData)` ensure the len is small than 255?